### PR TITLE
Cover index build and update with unit tests

### PR DESCRIPTION
**Pull request by zannis**

### CODE REVIEW

File: src/index.rs [hunk 1]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+                .expect("Should be fine");
+
+            let actual =
+                std::panic::catch_unwind(|| ResourceIndex::build(path.clone()));
+            assert!(actual.is_err());
+        })
+    }
+
+    #[test]
+    fn discover_paths_should_not_walk_on_invalid_path() {
```
Comment (kirillt) [inline src/index.rs:596]: Actually, it seems that we need to provide user app with clear error in this case.

**Review by kirillt — COMMENTED** [review]
Thanks, good job. Minor comments and questions only.

File: src/id.rs [hunk 2]
```diff
@@ -7,8 +7,8 @@ use log;
 
 #[derive(Eq, PartialEq, Hash, Clone, Debug)]
 pub struct ResourceId {
-    file_size: u64,
-    crc32: u32,
+    pub(crate) file_size: u64,
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/id.rs:10]: Difference is access to these properties will be available only throughout the crate (by tests for example). If these are to be accessed by other software consuming the library, they could be switched to public.

File: src/index.rs [hunk 3]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+    const CRC32_1: u32 = 3817498742;
+    const CRC32_2: u32 = 1804055020;
+
+    fn get_temp_dir() -> PathBuf {
+        create_dir_at(std::env::temp_dir())
+    }
+
+    fn create_dir_at(path: PathBuf) -> PathBuf {
+        let mut dir_path = path.clone();
+        dir_path.push(Uuid::new_v4().to_string());
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:298]: So the `std::env::temp_dir()` on mac is a static location, similar to `/tmp` on Linux. This  function does not create a new temp dir, so I append one to it everytime to work on a clean state and not deleting any temp files by mistake.

File: src/index.rs [hunk 4]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+        name: Option<&str>,
+    ) -> File {
+        let mut file_path = path.clone();
+        if let Some(file_name) = name {
+            file_path.push(file_name);
+        } else {
+            file_path.push(Uuid::new_v4().to_string());
+        }
+        let file = File::create(file_path).expect("Could not create temp file");
+        file.set_len(size.unwrap_or(0))
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:315]: I think not, file size is a u64 so, all these values would essentially have the same crc32 as they all resolve to 0.

File: src/index.rs [hunk 5]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+            assert_eq!(actual.ids.len(), 0);
+            assert_eq!(actual.collisions.len(), 0);
+            assert_eq!(actual.size(), 0);
+            assert_eq!(update.deleted.len(), 1);
+            assert_eq!(update.added.len(), 0);
+        })
+    }
+
+    #[test]
+    fn should_not_update_index_on_files_without_permissions() {
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:502]: I just made this one to be able to cover one of the errors. But there could be a valid case where some file can only be read by root and the user using the library has no read access to it.

File: src/index.rs [hunk 6]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+            assert_eq!(actual.size(), 2);
+
+            file.set_permissions(Permissions::from_mode(0o222))
+                .expect("Should be fine");
+
+            // let mut name_from = path.clone();
+            // name_from.push(FILE_NAME_2);
+            // let mut name_to = path.clone();
+            // path.clone().push(FILE_NAME_3);
+            // std::fs::rename(name_from, name_to);
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:524]: Correct

File: src/index.rs [hunk 7]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+                crc32: CRC32_1,
+            }));
+            assert!(actual.ids.contains(&ResourceId {
+                file_size: FILE_SIZE_2,
+                crc32: CRC32_2,
+            }));
+            assert_eq!(actual.collisions.len(), 0);
+            assert_eq!(actual.size(), 2);
+            assert_eq!(update.deleted.len(), 0);
+            assert_eq!(update.added.len(), 1);
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:470]: Ids are verified on lines 459-466.

File: src/index.rs [hunk 8]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+            let update = actual
+                .update()
+                .expect("Should update index successfully");
+
+            assert_eq!(actual.root, path.clone());
+            assert_eq!(actual.path2meta.len(), 0);
+            assert_eq!(actual.ids.len(), 0);
+            assert_eq!(actual.collisions.len(), 0);
+            assert_eq!(actual.size(), 0);
+            assert_eq!(update.deleted.len(), 1);
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:496]: Sure

File: src/index.rs [hunk 9]
```diff
@@ -268,3 +268,337 @@ fn is_hidden(entry: &DirEntry) -> bool {
...
+                crc32: CRC32_1,
+            }));
+            assert_eq!(actual.collisions.len(), 1);
+            assert_eq!(actual.size(), 2);
+        })
+    }
+
+    // resource index update
+
+    /// TODO: this case is probably not working correctly
```
[the remark this answers is not in the document]
Comment (zannis) [reply src/index.rs:378]: Sure, just added this as a note. Would appreciate some feedback on what would be the correct behavior here.

File: src/index.rs (hunk 1 above)
Comment (zannis) [reply src/index.rs:613]: Sure, let's keep this in mind for now and address in a followup issue as it probably involves code change as well.

File: src/index.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kirillt) [reply src/index.rs:299]: Oh, I got it. I thought it is equivalent of `mktemp` in Bash.

File: src/index.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kirillt) [reply src/index.rs:525]: Agree, it's good.

**Comment by kirillt** [comment]
`TODO` comments should be separated into new branch with creation of relevant issue and continuing the work in separate branch. The rest can be merged.

**Pull request merged**